packages/reactive-element/src/std-decorators/property.ts

```diff
@@ -0,0 +1,115 @@
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+/*
+ * IMPORTANT: For compatibility with tsickle and the Closure JS compiler, all
+ * property decorators (but not class decorators) in this file that have
+ * an @ExportDecoratedItems annotation must be defined as a regular function,
+ * not an arrow function.
+ */
+import {
+  PropertyDeclaration,
+  ReactiveElement,
+  defaultConverter,
+  notEqual,
+  propertyMetadata,
+} from '../reactive-element.js';
+
+// Overloads for property decorator so that TypeScript can infer the correct
+// return type when a decorator is used as an accessor decorator or a setter
+// decorator.
+export type PropertyDecorator = {
+  // accessor decorator signature
+  <C extends ReactiveElement, V>(
+    target: ClassAccessorDecoratorTarget<C, V>,
+    context: ClassAccessorDecoratorContext<C, V>
+  ): ClassAccessorDecoratorResult<C, V>;
+
+  // setter decorator signature
+  <C extends ReactiveElement, V>(
+    target: (value: V) => void,
+    context: ClassSetterDecoratorContext<C, V>
+  ): (this: C, value: V) => void;
+};
+
+// TODO (justinfagnani): export from ReactiveElement?
+// It actually makes sense to have this default defined with the decorator, so
+// that different decirators could have different defaults.
+const defaultPropertyDeclaration: PropertyDeclaration = {
+  attribute: true,
+  type: String,
+  converter: defaultConverter,
+  reflect: false,
+  hasChanged: notEqual,
+};
+
+/**
+ * Wraps a class accessor or setter so that `requestUpdate()` is called with the
+ * property name and old value when the accessor is set.
+ */
+export const property = (
+  options: PropertyDeclaration = defaultPropertyDeclaration
+): PropertyDecorator =>
+  (<C extends ReactiveElement, V>(
+    target: ClassAccessorDecoratorTarget<C, V> | ((value: V) => void),
+    context:
+      | ClassAccessorDecoratorContext<C, V>
+      | ClassSetterDecoratorContext<C, V>
+  ): ClassAccessorDecoratorResult<C, V> | ((this: C, value: V) => void) => {
+    const {kind, metadata} = context;
+
+    // Store the property options
+    let properties = propertyMetadata.get(metadata);
+    if (properties === undefined) {
+      propertyMetadata.set(metadata, (properties = new Map()));
+    }
+    properties.set(context.name, options);
+
+    if (kind === 'accessor') {
+      // Standard decorators cannot dynamically modify the class, so we can't
+      // replace a field with accessors. The user must use the new `accessor`
+      // keyword instead.
+      const {name} = context;
+      return {
+        get(this: C) {
+          return (target as ClassAccessorDecoratorTarget<C, V>).get.call(this);
+        },
+        set(this: C, v: V) {
+          const oldValue = (
+            target as ClassAccessorDecoratorTarget<C, V>
+          ).get.call(this);
+          (target as ClassAccessorDecoratorTarget<C, V>).set.call(this, v);
+          this.requestUpdate(name, oldValue, options);
+        },
+        init(this: C, v: V): V {
```

Comment:
We need a requestUpdate() call in here to add the property to the first update's changeProperties

Reply:
I'm concerned about the perf implications and complexity of using `queueMicrotask` here, especially since there's already an initial call to `requestUpdate`. 

Instead, it seems like populating `changedProperties` with the defaults on first rendered can just be handled in the first call to `update`.

Reply:
Directly populating changeProperties now